`FnPtr` blanket impls cause "multiple applicable items in scope" for unrelated types.

*Note: the below reduction originally came from a `use num_traits::Float;` + `0.0.max(x)` **in the wild**.*

I haven't found a way to demonstrate this outside of the "float inference vars + `Ord` method" combo, but:
```rust
trait MyCmp {
    fn cmp(&self) {}
}
impl MyCmp for f32 {}

fn main() {
    0.0.cmp();
}
```
works on stable (`1.68.2`) and beta (`1.69.0-beta.6`), and errors on nightly with:
```
error[E0034]: multiple applicable items in scope
 --> src/main.rs:7:9
  |
7 |     0.0.cmp();
  |         ^^^ multiple `cmp` found
  |
note: candidate #1 is defined in an impl of the trait `MyCmp` for the type `f32`
 --> src/main.rs:2:5
  |
2 |     fn cmp(&self) {}
  |     ^^^^^^^^^^^^^
  = note: candidate #2 is defined in an impl of the trait `Ord` for the type `F`
note: candidate #3 is defined in the trait `Iterator`
 --> /rustc/3a8a131e9509c478ece1c58fe0ea2d49463d2300/library/core/src/iter/traits/iterator.rs:3533:5
help: disambiguate the method for candidate #1
  |
7 |     MyCmp::cmp(&0.0);
  |     ~~~~~~~~~~~~~~~~
help: disambiguate the method for candidate #2
  |
7 |     Ord::cmp(&0.0);
  |     ~~~~~~~~~~~~~~
help: disambiguate the method for candidate #3
  |
7 |     Iterator::cmp(0.0);
  |     ~~~~~~~~~~~~~~~~~~
```

Even though it doesn't show the source for candidate `#2`, I'm almost certain it's this:
https://github.com/rust-lang/rust/blob/f13ccefb5b35af3dbb40a7bbdd64a3216d949664/library/core/src/ptr/mod.rs#L2064-L2070
from (cc @lcnr @oli-obk):
* https://github.com/rust-lang/rust/pull/99531
* https://github.com/rust-lang/rust/pull/108080

On the original PR I left comments like https://github.com/rust-lang/rust/pull/99531#issuecomment-1194688090, trying to rectify the issues with the original approach, but they didn't go anywhere and the version that one in the end is the (arguably) more fragile original one, that uses blanket `impl`s and works around their unwanted interactions.

It's possible in this case all it takes is checking for "float inference vars" and making sure they can never be considered to implement `FnPtr`, but I'm worried there might be other edge cases lurking in the shadows.

Also, did both of those PRs skip crater? I feel like this shouldn't be that hard to hit in the wild.

--- a/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs
+++ b/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs
@@ -998,8 +998,14 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             | ty::Alias(..)
             | ty::Param(..)
             | ty::Bound(..)
-            | ty::Error(_) => {}
-            ty::Infer(_) => {
+            | ty::Error(_)
+            | ty::Infer(
+                ty::InferTy::IntVar(_)
+                | ty::InferTy::FloatVar(_)
+                | ty::InferTy::FreshIntTy(_)
+                | ty::InferTy::FreshFloatTy(_),
+            ) => {}
+            ty::Infer(ty::InferTy::TyVar(_) | ty::InferTy::FreshTy(_)) => {
                 candidates.ambiguous = true;
             }
         }

--- a/compiler/rustc_trait_selection/src/traits/select/mod.rs
+++ b/compiler/rustc_trait_selection/src/traits/select/mod.rs
@@ -177,14 +177,14 @@ struct TraitObligationStack<'prev, 'tcx> {
 }
 
 struct SelectionCandidateSet<'tcx> {
-    // A list of candidates that definitely apply to the current
-    // obligation (meaning: types unify).
+    /// A list of candidates that definitely apply to the current
+    /// obligation (meaning: types unify).
     vec: Vec<SelectionCandidate<'tcx>>,
 
-    // If `true`, then there were candidates that might or might
-    // not have applied, but we couldn't tell. This occurs when some
-    // of the input types are type variables, in which case there are
-    // various "builtin" rules that might or might not trigger.
+    /// If `true`, then there were candidates that might or might
+    /// not have applied, but we couldn't tell. This occurs when some
+    /// of the input types are type variables, in which case there are
+    /// various "builtin" rules that might or might not trigger.
     ambiguous: bool,
 }
 

--- a/tests/ui/fn/fn-ptr-trait-int-float-infer-var.rs
+++ b/tests/ui/fn/fn-ptr-trait-int-float-infer-var.rs
@@ -1,0 +1,10 @@
+// check-pass
+trait MyCmp {
+    fn cmp(&self) {}
+}
+impl MyCmp for f32 {}
+
+fn main() {
+    // Ensure that `impl<F: FnPtr> Ord for F` is never considered for int and float infer vars.
+    0.0.cmp();
+}

